keps/0031-namespace-management.md

```diff
@@ -0,0 +1,152 @@
+
+## Proposals
+
+Prior to making KEP `provisional` we will chose a namespace creation proposal to move forward with and move the 2nd proposal to alternative.  For now, we have 2 competing namespace creation proposals.  
+In addition there is a proposal for the use of multiple namespaces.
+
+### Proposal 1: No namespace creation
+
+All namespace creation is managed outside KUDO. Users such as D2iQ would need to find or build automation for the creation of their namespaces. This REQUIRES that KUDO honor the deployment of resources to the "default" namespace, or to the namespace defined by `--namespace` during install.  
+
+KUDO should further require that NO namespaces be allow in manifests for resources deployed during "install".  The concept of defining manifest which land in multiple namespaces is defined in a latter proposal and is a separate concern. Without multi-namespace support, all resources are install to one namespace and namespace must be created prior to installation and that namespace must be specified during installation.
+
+### Proposal 2: Namespace Creation
+
+KUDO should support the creation of a namespace if it does not exist.  Further KUDO will add a configuration for namespace such that at time of namespace creation it is created with the metadata provided. To this end a manifest file for namespace is to be supported in the template folder:
+
+```
+apiVersion: v1
+kind: Namespace
+metadata:
+  labels:
+    ca.istio.io/override: "true"
+    istio-injection: enabled
+    katib-metricscollector-injection: enabled
+  name: {{ .Namespace }}
+```
+
+The template supports the implicit `.Namespace` parameter (which already exists).  The `operator.yaml` file is to will be extended to support `namespaceManifest`.
+
+```
+# operator.yaml
+name: "kubeflow"
+operatorVersion: "0.2.0"
+kudoVersion: "0.10.1"
+kubernetesVersion: 1.15.0
+appVersion: 1.0.0
+namespaceManifest: templates/namespace.yaml
+```
+
+The follow rules apply:
+
+1. Missing `namespaceManifest` means that there is no metadata (creation still applies if warranted)
+1. Missing `name: {{ .Namespace }}` is a failure (and will be validated by package verifier)
+1. No support for static namespace, it must be templated or the manifest should not exist
+1. If no `namespaceManifest` and no `--namespace` the operator installs to "default"
+1. If `--namespace` provided and namespace doesn't exist, KUDO creates it
+1. If KUDO creates namespace, if there is a `namespaceManifest` it uses it to create namespace otherwise does a simple namespace creation.
+1. If KUDO has `namespaceManifest` the check on the namespace is a deepcopy (mean it will ensure that the labels and annotations exists and match) if not it overwrites
+1. if a namespace is created KUDO waits until the namespace is created to move forward in a plan.
+1. if a namespace is needed to be created, it is detected early in the process prior to the deployment plan or as a first step.
+
```

Comment:
I feel that the rules around namespaces are too complex. I'd rather make namespace creation explicit and **require** the `namespaceManifest` for it.  Wdyt about the following simplification:

1. If KUDO should create the namespace, a `namespaceManifest` **should be present**. This also implies that `--namespace` alone is **not enough** to create a namespace automatically
2. `--namespace` is used only to populate the `name: {{ .Namespace }}`, otherwise whatever the user has configured in her setting is taken
3. If defined namespace already exists, it's an **error and we abort the installation**. We don't try to merge/adjust labels or annotations as it has unforeseeable consequences.

Reply:
I agree with the complexity :(   I would like to get your take on the latest change.. which is to be explicit about create with `--apply-ns`